consensus/istanbul/backend/engine.go

```diff
@@ -175,20 +177,112 @@ func (sb *Backend) verifyCascadingFields(chain consensus.ChainReader, header *ty
 func (sb *Backend) VerifyHeaders(chain consensus.ChainReader, headers []*types.Header, seals []bool) (chan<- struct{}, <-chan error) {
 	abort := make(chan struct{})
 	results := make(chan error, len(headers))
-	go func() {
-		for i, header := range headers {
-			err := sb.verifyHeader(chain, header, headers[:i])
 
-			select {
-			case <-abort:
-				return
-			case results <- err:
+	// If we are in "lightest" mode, we do batch verification of all BLS signatures
+	// in the headers, so there is no need to iterate over the blocks
+	if chain.Config().FullHeaderChainAvailable {
+		go func() {
+			for i, header := range headers {
+				err := sb.verifyHeader(chain, header, headers[:i])
+
+				select {
+				case <-abort:
+					return
+				case results <- err:
+				}
 			}
-		}
-	}()
+		}()
+	} else {
+		go func() {
+			err := sb.batchVerifyHeaders(chain, headers)
```

Comment:
Should we batch verify headers even when `FullHeaderChainAvailable` == true? In those cases for the most part the validator set will be the same for every header in the batch

Reply:
In this case we can even do better, tbh.

Reply:
Cool, it sounds like it would be both more efficient *and* simplify the code, so I suggest we go for it